### ISSUE TITLE
Possible Race Condition between Compiling and Linking

### ISSUE THREAD

**Issue by missmah** [fastbuild/fastbuild#670]
I wrote a tool to make our `All` builds faster by "pretending to link" a bunch of targets we rarely test with when a `#define` is set in `fbuild.bff` to enable this hacky mode.

The idea here is that we pretty much always want to compile all the code on all the platforms to check for compiler errors. OTOH, many of our platforms rarely have linker errors, but linking generally takes longer than compiling, and when iterating locally, if we can skip most of the linking for the platforms we won't be testing with during a given work session, we can save a lot of wasted energy/time.

So, I built this `IvrLinkerEmulatorTool`. It understands how to parse command line arguments for all the linkers we use (`MSVC`, `Clang`, `Emscripten`, etc.). And internally, it just does a file copy from the first input object file to the output file. If the output file already exists, it just touches it, rather than doing the copy. But it will fail if the input file is still locked for writing by say a `compiler` process.

When this hacky mode is enabled, building `All`, I get sporadic "linker failures" on `Android-ARM64-Shipping` targets.

When I build `All-Android-ARM64-Shipping`, I get no failures at all.

This led me to believe this may be a race condition in `fastbuild` between the compiler and the linker (that the linker starts before the compiler has released the file at the `OS/kernel` level) - and that this race is more likely to result in a failure when the system is heavily loaded (e.g. building `All`) than when it's lightly loaded (e.g. building one single `All-Platform-Configuration`).

This could actually be a `Win32 Kernel` bug, but statistically, `fastbuild` seems more likely to be the culprit.

So, I added some code to `IvrLinkerEmulatorTool` to test my hypothesis:

```
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// FastBuild appears to have a race condition.
// FastBuild All with all cores fully loaded can have CopySingleFile below fail with 0x20 (File is locked by another Process).
// The file which is locked is the .obj file (SrcInputFile).
// What appears to happen is that FastBuild starts the linker process before the previous 
// compiler process has fully finished unlocking the file (at the kernel level?).
// This of course also could be a Win32 Kernel bug, though that seems statistically less likely.
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
// So, to mitigate the race condition, we added the code below. We give the file copy iNumRetries chances.
// We sleep iNumMillisecondsToSleepBetweenRetries milliseconds between each copy attempt.
// With iNumRetries set to 5, and iNumMillisecondsToSleepBetweenRetries = 100, the race condition is fully mitigated for
// my local system with 72 logical cores and using our codebase with FastBuild All
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
bool bCopiedFileOk = false;
const int iNumRetries = 5;
const int iNumMillisecondsToSleepBetweenRetries = 100;
for( int i = 0; i < iNumRetries; ++i )
{
	const bool bCopiedFileOkThisTime = m_pFileSystem->CopySingleFile( i_asciiDstOutputFile, i_asciiSrcInputFile, my_eFileCopyOptions );
	if( bCopiedFileOkThisTime )
	{
		bCopiedFileOk = true;
		break;
	}
	IvrSysSleep( iNumMillisecondsToSleepBetweenRetries );
}
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```

And `All` now succeeds 100% of the time.

Realistically, given that this race condition likely never actually manifests itself in real workloads (as it requires such an unreasonably *fast linker*), this is probably not worth fixing. :)

Particularly because fixing it would probably slow down all builds at the expense of fixing a problem which, again, probably only ever manifests itself when you're using a "cheater linker" like what I'm doing. And the fix inside the cheater linker is only a handful of lines of code, and probably even in that situation, more efficient than fixing the original race condition.

But, I thought I should probably document this for your knowledge @ffulin, and in-case anyone else runs into the situation I hit by experimenting with artificial linkers.

------------------------------------------------------------------------------------------------------

P.S. On that subject, the cheater linker thing is such a success that I'm thinking of making the fake linker do linker-level caching so that it would have a mode where it wouldn't actually be a cheater linker anymore. But, the ideal way to do that caching would be to be able to know what the hashes of the pre-processed source used to build the object file were, and the hash of the linker command line arguments, and if all of those hashes matched in the cache, then you could safely pull the output file out of the cache.

@ffulin do you think it would be possible to add an API to get those pre-processed source hashes as strings that could be passed to the "linker"? Basically like `%4` = `%1` but hashes instead of filenames (in the same order as the filenames in `%1`). If this is something you'd be interested in supporting, I could file a separate issue for this feature proposal, so we could discuss the best way to do the implementation, and who should do it (me? you? someone else?).

**Comment by missmah** [comment]
More "obvious" proof of the race condition:

Link times using `IvrLinkerEmulatorTool.exe` for both the linker and librarian on a project:

# 10ms Sleeps between Retries adds only about `10ms` (one retry) to the `.dll` link:

`const int iNumRetries =25;`
`const int iNumMillisecondsToSleepBetweenRetries = 10;`
```
--- Most Expensive ----------------------------------------------
Time (s)  Name:
0.065     D:\dev\Build\Output\x64-Shipping\PicoMath\IvrGfxMath\IvrGfxMath-Unity_1.obj
0.039     D:\dev\Build\Output\x64-Shipping\PicoMath\IvrGfxMath\IvrGfxMath.dll
0.030     D:\dev\Build\Output\x64-Shipping\PicoMath\IvrGfxMath\IvrGfxMath.lib
```

# 1000ms Sleeps between Retries adds `1 second` to the `.dll` link:

`const int iNumRetrie = 25;`
`const int iNumMillisecondsToSleepBetweenRetries = 1000;`
```
1.029     D:\dev\Build\Output\x64-Shipping\PicoMath\IvrGfxMath\IvrGfxMath.dll
0.065     D:\dev\Build\Output\x64-Shipping\PicoMath\IvrGfxMath\IvrGfxMath-Unity_1.obj
0.029     D:\dev\Build\Output\x64-Shipping\PicoMath\IvrGfxMath\IvrGfxMath.lib
```

**Comment by missmah** [comment]
# With Cache enabled and also skip linking enabled: `1.592s`

1. It feels like this should be much closer to zero. The link steps are taking `350-400ms`, while without any retries, they should take about `29-30ms` based on previous testing. This is more than `10X` overhead for `"copy retries"` -- would it be possible to have a mode where we skip actually copying the output `object file` from the `cache` to the output folder on a hit where the `timestamp` hasn't changed on disk? I know this should not be the default, because there are situations where it isn't safe -- but, it would likely reduce the filesystem contention being seen here by quite a bit, and get us much closer to `0s` for a fully `cached` + `skip-link` build (with the goal here to have near `zero` overhead on objects that don't need to be rebuilt by an actual compiler).

```
--- Most Expensive ----------------------------------------------
Time (s)  Name:
0.664     D:\dev\Build\Output\x64-Development\NanoGSTDecompressor\glfw\glfw-x64-Development-Unity_1.obj
0.427     D:\dev\Build\Output\CSharp-Development\GstDebugViewer\GstDebugViewer_CSharpApp\GstDebugViewer_CSharpApp.exe
0.394     D:\dev\Build\Output\x64-Development\MicroAudioAPI\MicroAudioAPI\MicroAudioAPI.dll
0.391     D:\dev\Build\Output\x64-Development\NanoIvrDevUtilities\NanoIvrDevUtilities\NanoIvrDevUtilities.dll
0.383     D:\dev\Build\Output\x64-Development\MicroShaderCompilerAPI\MicroShaderCompilerAPI\MicroShaderCompilerAPI.dll
0.383     D:\dev\Build\Output\x64-Development\InsightfulUnityPlugin\InsightfulUnityPlugin\InsightfulUnityPlugin.dll
0.382     D:\dev\Build\Output\x64-Development\PicoStringPool\PicoStringPool\PicoStringPool.dll
0.380     D:\dev\Build\Output\x64-Development\MicroBasicForwardRenderer\MicroBasicForwardRenderer\MicroBasicForwardRenderer.dll
0.379     D:\dev\Build\Output\x64-Development\NanoOpenGL_GL3\NanoOpenGL_GL3\NanoOpenGL_GL3.dll
0.378     D:\dev\Build\Output\x64-Development\UtilityCore\UtilityCore_ExampleApp\UtilityCore_ExampleApp.exe
0.376     D:\dev\Build\Output\x64-Development\GigaSerializer\GigaSerializer_ExampleApp\GigaSerializer_ExampleApp.exe
0.375     D:\dev\Build\Output\x64-Development\MicroEventRecorder\MicroEventRecorder\MicroEventRecorder.dll
0.375     D:\dev\Build\Output\x64-Development\MicroProcessProfileDB\MicroProcessProfileDB_ExampleApp\MicroProcessProfileDB_ExampleApp.exe
0.374     D:\dev\Build\Output\x64-Development\MicroTranscodeAPI\MicroTranscodeAPI\MicroTranscodeAPI.dll
0.374     D:\dev\Build\Output\x64-Development\MemTrace\MemTrace_ExampleApp\MemTrace_ExampleApp.exe
0.373     D:\dev\Build\Output\x64-Development\NanoNet\NanoNetServerTest\NanoNetServerTest.exe
0.373     D:\dev\Build\Output\x64-Development\NanoCore\NanoCore\NanoCore.dll
0.373     D:\dev\Build\Output\x64-Development\MicroNet\MicroNet\MicroNet.dll
0.373     D:\dev\Build\Output\x64-Development\NanoNet\IVRStandaloneConnectionManager\IVRStandaloneConnectionManager.exe
0.372     D:\dev\Build\Output\x64-Development\MemoryCore\MemoryCore_ExampleApp\MemoryCore_ExampleApp.exe

--- Summary -----------------------------------------------------
                                 /----- Cache -----\
Build:          Seen    Built   Hit     Miss    Store   CPU
 - CopyFile   : 336     336     -       -       -       16.342s
 - Directory  : 153     153     -       -       -       0.069s
 - File       : 3273    347     -       -       -       0.002s
 - Library    : 14      14      -       -       -       2.396s
 - Object     : 157     1       156     0       0       18.018s
 - Alias      : 1       1       -       -       -       0.000s
 - Exe        : 75      75      -       -       -       20.926s
 - Unity      : 146     146     -       -       -       4.890s
 - C#         : 4       4       -       -       -       0.962s
 - Compiler   : 3       3       -       -       -       0.000s
 - DLL        : 53      53      -       -       -       11.794s
 - ObjectList : 143     143     -       -       -       0.000s
 - TextFile   : 4       4       -       -       -       0.151s
Cache:
 - Hits       : 156 (100.0 %)
 - Misses     : 0
 - Stores     : 0
Time:
 - Real       : 1.384s
 - Local CPU  : 1m 15.550s (54.6:1)
 - Remote CPU : 0.000s (0.0:1)
-----------------------------------------------------------------
FBuild: OK: All-x64-Development
Time: 1.592s
```

**Comment by ffulin** [comment]
> Possible race condition

It's prettyunlikely there is a race condition in FASTBuild, because the compilation and linking are enitrely separate tasks, with no interaction between them. The linking cannot start until the compiler process has exited and another sweep of the dependency graph occurs. Such a race condition would have to exist due to a really fundamental bug in the core of the build engine and would manifest all over the place. Nothing can ever be dismissed entirely of course.

My money would be on interference from security software. On Windows, it's very common that security software interferes withe file access, particularly when:
 - one process creates a file
 - another process immediately consumes that file

I've seen cases where Antivirus will see the first write and try to "scan" the file, preventing the second application opening it.  The specific case I saw was when FASTBuild would write the preprocessed output to disk and invoke the compiler, which would then fail to open the file. FASTBuild has a (Windows only) work-around for this (WorkAroundForWindowsFilePermissionProblem). This problem was far more frequently seen under heavy load and pointed to flaws in the security software/AV implementation (not using OpLocks for example)

If this is happening in your case, it's possible that the real linker takes significantly longer to start before attempting to open the file, avoiding the race-condition, while your light-weight app might open the file much more quickly.  It's also possible that the MSVC linker (or other linkers too) have retry logic to work around this problem. Finally, you might be opening the file with different sharing permissions.

Things you can try to better understand this problem:
 - try on another OS - is this problem Windows-specific?
 - try disabling any security software you have (if you can) or try excluding directories or file types involved in the problem
 - try using something like [ProcessMonitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) to understand the file access pattern across the different involved processes

Assuming it is security software related and Windows specific, a work-around may be your only option. It seems most appropriate for such a work-around to be in your "fake linker" because:
 - it's the only thing impacted that we know of
 - you can only retry only failure, paying the cost only when there is a problem (instead of FASTBuild speculatively paying a cost all the time (for every file that's an input into the link) as it does with the other work-around)

If we get to the point where we definitively understand the problem and could apply a low-cost work around to FASTBuild, I'd be open to that though.

**Caching of linking**

FASTBuild doesn't have all the necessary information needed to cache linking - that's why it doesn't do it :) Gaining the necessary information for safe caching is far more complex that the .obj caching case.

Caching linking results requires hashes of at least the following:
a) the object files build by FASTBuild (if fed directly to the link) (the hash of the source might be used as a proxy for this)
b) libraries built by FASTBuild (if fed directly to the link)
c) existing object files or libraries on disk (3rd party libraries, the system/os SDK etc)
d) other things (like manifests, resources or other things that impact the link
e) external/3rd party libraries
f) linker toolchain (link.exe and associated dlls and helpers)
g) the linker command line

FASTBuild doesn't have most of these available. Obtaining hashes for these things is possible, but would add non-trivial overhead to the build. This would make what I would consider the "common case" (where the link will be unique) much slower, to benefit the case where you are rebuilding something with literally nothing different. I could perhaps see a codebase that has *lots* of dlls, most of which remain static, potentially seeing a benefit though.

If we went and added code to FASTBuild to obtain the hashes for the things it needs, there wouldn't be much point passing them to your fake linker, because at that point we can just have FASTBuild hit the cache.

**Comment by ffulin** [comment]
For the core issue in this ticket (possible race condition between compiling and linking) I'm reasonably confident there is no issue on the FASTBuild side and so there is nothing actionable. With that in mind, I'm going to close this ticket to help prune the backlog. We can re-open or create a new issue if new information comes to light.
